### PR TITLE
8293562: KeepAliveCache Blocks Threads while Closing Connections

### PR DESCRIPTION
**Pull request by PoojaDP-23**
fix: https://bugs.openjdk.org/browse/JDK-8293562

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293562](https://bugs.openjdk.org/browse/JDK-8293562): KeepAliveCache Blocks Threads while Closing Connections (**Bug** - P4)

### Reviewers
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk11u-dev.git pull/1890/head:pull/1890` \
`$ git checkout pull/1890`

Update a local copy of the PR: \
`$ git checkout pull/1890` \
`$ git pull https://git.openjdk.org/jdk11u-dev.git pull/1890/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1890`

View PR using the GUI difftool: \
`$ git pr show -t 1890`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk11u-dev/pull/1890.diff">https://git.openjdk.org/jdk11u-dev/pull/1890.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk11u-dev/pull/1890#issuecomment-1557112259)

### CODE REVIEW

**Comment by jerboaa** [comment]
Seems a duplicate of #1825

**Comment by alograg** [comment]
Hi @PoojaDP-23,
I was working on the backporting of the fix. #1825 
But after the @jerboaa reviews, I realized that your code is closer to the original fix commit than the mine.
I offer you all my support, I remain at your disposal.

**Comment by jerboaa** [comment]
@PoojaDP-23 Please use `Backport 03f25a9c6924430ec4063b801b2b6ca55b9067c9` as the PR title so that bots recognize it as a backport. Also, please enable Github Action testing on your fork.

**Comment by PoojaDP-23** [comment]
> Hi @PoojaDP-23, I was working on the backporting of the fix. #1825 But after the @jerboaa reviews, I realized that your code is closer to the original fix commit than the mine. I offer you all my support, I remain at your disposal.

Thanks

**Comment by jerboaa** [comment]
@PoojaDP-23 Please also merge latest `master` tree into the PR branch so as to trigger GHA checks for this PR. Thanks!

**Comment by PoojaDP-23** [comment]
@jerboaa - Yes, I have enabled  Github Action testing on my fork

**Comment by PoojaDP-23** [comment]
@jerboaa - I have performed below steps to merge latest master into PR branch. Could you please take a look?
```
git checkout master
git pull https://github.com/openjdk/jdk17u-dev master
git checkout <Your_BRANCH_NAME>
git merge master
git push origin HEAD

**Comment by alograg** [comment]
Hi @jerboaa 
Can you do the review ?
Thrnaks,

File: src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java [hunk 1]
```diff
@@ -268,8 +290,8 @@ private void readObject(ObjectInputStream stream)
     }
 }
 
-/* FILO order for recycling HttpClients, should run in a thread
- * to time them out.  If > maxConns are in use, block.
+ /* LIFO order for reusing HttpClients. Most recent entries at the front.
```
Comment (jerboaa) [inline src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java:293]: Nit: Extraneous space in front of `/*`.

File: test/jdk/sun/net/www/http/KeepAliveCache/B8293562.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+                    } else {
+                        result.completeExceptionally(new RuntimeException(
+                                "Wait for second request timed out"));
+                    }
+                } catch (InterruptedException e) {
+                    result.completeExceptionally(new RuntimeException(
+                            "Wait for second request was interrupted"));
+                }
+            } else {
+                System.out.println("name of the thread" +threadName);
```
Comment (jerboaa) [inline test/jdk/sun/net/www/http/KeepAliveCache/B8293562.java:138]: This isn't part of the original change. Please remove.

File: test/jdk/sun/net/www/http/KeepAliveCache/B8293562.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+                result.completeExceptionally(new RuntimeException(
+                        "Close invoked from unexpected thread"));
+            }
+            System.out.println("Connection closed");
+            System.out.println(threadName);
```
Comment (jerboaa) [inline test/jdk/sun/net/www/http/KeepAliveCache/B8293562.java:143]: Same here. Extraneous debug `println()`?

**Review by jerboaa — COMMENTED** [review]
This seems OK to me. A few minor comments. What testing have you done?

**Comment by darinhoward** [comment]
The docs at https://www.oracle.com/java/technologies/javase/11-0-17-relnotes.html indicate this fix was done in 11.0.17.0.3.  I'm assuming this is in error based on this PR still being open.

**Comment by alograg** [comment]
Hi @darinhoward 
The bug is fixed in version JDK 11.0.17.0.3 (Oracle) not in [OpenJDK 11.0.17](https://bugs.openjdk.org/browse/JDK-8293562)

**Comment by jerboaa** [comment]
> The docs at https://www.oracle.com/java/technologies/javase/11-0-17-relnotes.html indicate this fix was done in 11.0.17.0.3. I'm assuming this is in error based on this PR still being open.

Yes, those are Oracle JDK release notes. This PR is for OpenJDK.

**Comment by PoojaDP-23** [comment]
@jerboaa - I have addressed review comments and I have enabled GHA actions to test and I have separately run the testcase "B8293562.java" with(passes) and without(fails) the fix to test these changes. Thanks

**Review by jerboaa — APPROVED** [review]

**Comment by PoojaDP-23** [comment]
@jerboaa - Could you please add the fix request to the jbs issue? Please let me know if there are any actions from my side. Thanks

**Comment by jerboaa** [comment]
> @jerboaa - Could you please add the fix request to the jbs issue? Please let me know if there are any actions from my side. Thanks

Done. Please mention what extra testing of the patch you've done here (other than the regtest). Thanks!

**Comment by GoeLin** [comment]
Hi @PoojaDP-23 
You should not integrate before you have the fix-yes label!!!

I assume JDK-8296438, which is listed as related issue, is a follow up.

**Comment by jerboaa** [comment]
Please check https://bugs.openjdk.org/browse/JDK-8296943 before this gets sponsored.

**Comment by PoojaDP-23** [comment]
@GoeLin - Apologies for commenting integrate. @jerboaa - I have enabled GHA actions and also I have ran the TIER1 and TIER2 tests and all are passing and I have ran the jtreg test [test/jdk/sun/net/www/http/KeepAliveCache/B8293562.java](https://github.com/openjdk/jdk11u-dev/pull/1890/files#diff-43af219708466772a71875d5d53dbfb340fe012b521788268716c242c668a834) which passes with this fix and fails without the fix
TIER: 1
Test summary
```
   TEST                                              TOTAL  PASS  FAIL ERROR
   jtreg:test/hotspot/jtreg:tier1                     1482  1482     0     0
   jtreg:test/jdk:tier1                               1881  1881     0     0
   jtreg:test/langtools:tier1                         3934  3934     0     0
   jtreg:test/nashorn:tier1                              0     0     0     0
   jtreg:test/jaxp:tier1                                 0     0     0     0
```

TEST SUCCESS


TIER: 2
Test summary

  ```
   TEST                                              TOTAL  PASS  FAIL ERROR
   jtreg:test/hotspot/jtreg:tier2                      325   325     0     0
   jtreg:test/jdk:tier2                               3468  3468     0     0
   jtreg:test/langtools:tier2                           11    11     0     0
   jtreg:test/nashorn:tier2                             36    36     0     0
   jtreg:test/jaxp:tier2                               440   440     0     0

```

TEST SUCCESS

**Comment by PoojaDP-23** [comment]
@jerboaa - I have checked https://bugs.openjdk.org/browse/JDK-8296943 and I have validated the test sun/net/www/http/HttpClient/MultiThreadTest.java with this fix and test is passing. Please take a look on below screenshot.

**Java-version:**
image




**Test-result**
image

**Comment by jerboaa** [comment]
OK, thanks!

**Pull request closed without merging**